src/stream/src/executor/project/materialized_exprs.rs

```diff
@@ -0,0 +1,216 @@
+        self.state_table.inner.init_epoch(first_epoch).await?;
+
+        #[for_await]
+        for msg in input {
+            let msg = msg?;
+            match msg {
+                Message::Chunk(input_chunk) => {
+                    let mut eval_visibility = BitmapBuilder::from(input_chunk.visibility().clone());
+                    for (i, op) in input_chunk.ops().iter().enumerate() {
+                        // hide deletions from expression evaluation
+                        match op {
+                            Op::Delete | Op::UpdateDelete => eval_visibility.set(i, false),
+                            _ => {}
+                        }
+                    }
+                    let eval_chunk = input_chunk
+                        .data_chunk()
+                        .with_visibility(eval_visibility.finish());
+
+                    let mut eval_result_arrs = Vec::with_capacity(self.exprs.len());
+                    for expr in &self.exprs {
+                        // for deletions, the evaluation result is NULL
+                        eval_result_arrs.push(expr.eval_infallible(&eval_chunk).await);
+                    }
+
+                    let mut eval_result_builders = eval_result_arrs
+                        .iter()
+                        .map(|arr| arr.create_builder(input_chunk.capacity()))
+                        .collect::<Vec<_>>();
+                    // now we need to replace the NULLs with the old evaluation results
+                    for (row_idx, row_op) in input_chunk.rows_with_holes().enumerate() {
+                        let Some((op, row)) = row_op else {
+                            // it's invisible in the input
+                            for builder in &mut eval_result_builders {
+                                builder.append_null();
+                            }
+                            continue;
+                        };
+
+                        match op {
+                            Op::Insert | Op::UpdateInsert => {
+                                // for insertions, append the evaluation results
+                                let output_row = row
+                                    .iter()
+                                    .chain(
+                                        eval_result_arrs
+                                            .iter()
+                                            .zip_eq_fast(&mut eval_result_builders)
+                                            .map(|(arr, builder)| {
+                                                let datum_ref =
+                                                    unsafe { arr.value_at_unchecked(row_idx) };
+                                                builder.append(datum_ref);
+                                                datum_ref
+                                            }),
+                                    )
+                                    .collect::<Vec<_>>();
```

Comment:
I suppose there's no need to collect a `Vec` if we split the `builder.append` step from constructing the state table row?